Source/FieldSolver/ImplicitSolvers/WarpXImplicitOps.cpp

```diff
@@ -0,0 +1,333 @@
+
+            setPosition(ip, xp, yp, zp);
+        });
+
+    }
+    }
+
+    }
+
+}
+
+void
+WarpX::FinishImplicitField( amrex::Vector<std::array< std::unique_ptr<amrex::MultiFab>, 3 > >& Field_fp,
+                      const amrex::Vector<std::array< std::unique_ptr<amrex::MultiFab>, 3 > >& Field_n,
+                      amrex::Real theta )
+{
+    using namespace amrex::literals;
+
+    for (int lev = 0; lev <= finest_level; ++lev) {
+
+#ifdef AMREX_USE_OMP
+#pragma omp parallel if (amrex::Gpu::notInLaunchRegion())
+#endif
+       for ( amrex::MFIter mfi(*Field_fp[lev][0], amrex::TilingIfNotGPU()); mfi.isValid(); ++mfi )
+        {
+
+            amrex::Array4<amrex::Real> const& Fx = Field_fp[lev][0]->array(mfi);
+            amrex::Array4<amrex::Real> const& Fy = Field_fp[lev][1]->array(mfi);
+            amrex::Array4<amrex::Real> const& Fz = Field_fp[lev][2]->array(mfi);
+
+            amrex::Array4<amrex::Real> const& Fx_n = Field_n[lev][0]->array(mfi);
+            amrex::Array4<amrex::Real> const& Fy_n = Field_n[lev][1]->array(mfi);
+            amrex::Array4<amrex::Real> const& Fz_n = Field_n[lev][2]->array(mfi);
+
+            amrex::Box const& tbx = mfi.tilebox(Field_n[lev][0]->ixType().toIntVect());
+            amrex::Box const& tby = mfi.tilebox(Field_n[lev][1]->ixType().toIntVect());
+            amrex::Box const& tbz = mfi.tilebox(Field_n[lev][2]->ixType().toIntVect());
+
+            const amrex::Real c0 = 1._rt/theta;
+            const amrex::Real c1 = 1._rt - c0;
+
+            amrex::ParallelFor(
+            tbx, ncomps, [=] AMREX_GPU_DEVICE (int i, int j, int k, int n)
+            {
+                Fx(i,j,k,n) = c0*Fx(i,j,k,n) + c1*Fx_n(i,j,k,n);
+            },
+            tby, ncomps, [=] AMREX_GPU_DEVICE (int i, int j, int k, int n)
+            {
+                Fy(i,j,k,n) = c0*Fy(i,j,k,n) + c1*Fy_n(i,j,k,n);
+            },
+            tbz, ncomps, [=] AMREX_GPU_DEVICE (int i, int j, int k, int n)
+            {
+                Fz(i,j,k,n) = c0*Fz(i,j,k,n) + c1*Fz_n(i,j,k,n);
+            });
+        }
+    }
+}
+
+void
+WarpX::ImplicitComputeRHSE (amrex::Real a_dt, WarpXSolverVec&  a_Erhs_vec)
+{
+    for (int lev = 0; lev <= finest_level; ++lev)
+    {
+        ImplicitComputeRHSE(lev, a_dt, a_Erhs_vec);
+    }
+}
+
+void
+WarpX::ImplicitComputeRHSE (int lev, amrex::Real a_dt, WarpXSolverVec&  a_Erhs_vec)
+{
+    WARPX_PROFILE("WarpX::ImplicitComputeRHSE()");
+    ImplicitComputeRHSE(lev, PatchType::fine, a_dt, a_Erhs_vec);
+    if (lev > 0)
+    {
+        ImplicitComputeRHSE(lev, PatchType::coarse, a_dt, a_Erhs_vec);
+    }
+}
+
+void
+WarpX::ImplicitComputeRHSE (int lev, PatchType patch_type, amrex::Real a_dt, WarpXSolverVec&  a_Erhs_vec)
+{
+    // set RHS to zero value
+    a_Erhs_vec.getVec()[lev][0]->setVal(0.0);
+    a_Erhs_vec.getVec()[lev][1]->setVal(0.0);
+    a_Erhs_vec.getVec()[lev][2]->setVal(0.0);
+
+    // Compute Efield_rhs in regular cells by calling EvolveE
```

Comment:
```suggestion
    // Compute Efield_rhs in regular cells by calling EvolveE
    // Because `a_Erhs_vec` was set the zero in the code above, calling
    // `EvolveE` on `a_Erhs_vec` will result in `a_Erhs_vec` storing only 
    // the RHS of the update equation (i.e. c^2*dt*( curl(B^{n+theta}) - mu0*J^{n+1/2} ))
```

Reply:
done